Cabal/src/Distribution/PackageDescription/Check/Warning.hs

```diff
@@ -264,6 +298,441 @@ extractCheckExplantion (PackageDistSuspicious e) = e
 extractCheckExplantion (PackageDistSuspiciousWarn e) = e
 extractCheckExplantion (PackageDistInexcusable e) = e
 
+-- | Identifier for the speficic 'CheckExplanation'. This ensures `--ignore`
+-- can output a warning on unrecognised values.
```

Comment:
Maybe
```suggestion
-- can output a warning on unrecognised values.
-- NOTE: should be kept in sync with 'CheckExplanation'
```
or something like that.

To be honest, I'm not a fan of another data type with 137 constructors (and counting). But I don't have great ideas about it. Something fancier like `singletons` comes to mind...

1.  Keep `CheckExplanationID` as is.
2.  Change `CheckExplanation` to have one constructor with an extension point defined by a type family, e.g.

```
data CheckExplanation (id :: CheckExplanationID) =
  CE (Sing id) (CheckExplanationPayload id)

type family CheckExplanationPayload id where
  CheckExplanationPayload 'DuplicateSections = [UnqualComponentName]
  -- ... many more for every explanation with non-trivial payload
  CheckExplanationPayload id = () -- default payload is unit
```

Not sure it's the best way to skin this cat. Maybe someone with _working knowledge_ of Haskell beyond Haksell98 could suggest a better solution...

And if you change ChangeExplanation, you'll have to change a ton of code, I guess... So, I don't want to hold this PR hostage of this observation.

Reply:
This is an excellent suggestion. I have never actually used type familis, but they seem quite suited to the problem. Plan: merge this, then I will investigate how difficult it is to implement, report back on Matrix..

Reply:
Also: “keep in sync” invariant comes for free with `-WAll`.

Reply:
SGTM. Good to know about -Wall. Wouldn't hurt to spell it out in the docs anyway I think.